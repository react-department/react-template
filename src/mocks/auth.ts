export default {
  id: 1,
  token: 'token',
};
