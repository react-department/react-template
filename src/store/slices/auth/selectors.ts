import { createSelector } from '@reduxjs/toolkit';

import type { RootState } from '../../types/TStore';
import type IAuthResponse from './interfaces/IAuthResponse';

const selectAuth = (state: RootState): IAuthResponse => state.auth;

export const selectUserId = createSelector(selectAuth, (auth) => auth.id);
export const selectUserToken = createSelector(selectAuth, (auth) => auth.token);

export const selectCurrentUser = createSelector(
  (data: IAuthResponse | undefined) => data,
  (data: IAuthResponse | undefined, userId: IAuthResponse['id'] | undefined) => userId,
  (data, userId) => (userId ? data : undefined),
);
