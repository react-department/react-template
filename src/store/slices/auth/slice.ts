import { createSlice } from '@reduxjs/toolkit';

import dummyAuthApi from './apis/dummyAuth';

import type IAuthResponse from './interfaces/IAuthResponse';

export const initialState: IAuthResponse = {
  id: undefined,
  token: undefined,
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {
    logout: () => initialState,
  },
  extraReducers: (builder): void => {
    builder.addMatcher(
      dummyAuthApi.endpoints.login.matchFulfilled,
      (state, { payload }): void => {
        state.id = payload.id;
        state.token = payload.token;
      },
    );
  },
});

export const { logout } = userSlice.actions;

export default userSlice.reducer;
