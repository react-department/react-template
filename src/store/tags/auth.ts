const prefix = 'AUTH/';
const AUTHENTICATED_USER = `${prefix}AUTHENTICATED_USER`;

export default AUTHENTICATED_USER;
