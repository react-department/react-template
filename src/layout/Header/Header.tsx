import { useTranslation } from 'react-i18next';
import { Link } from 'react-router-dom';

import LinkButton from '../../components/LinkButton/LinkButton';
import Loader from '../../components/Loader/Loader';
import Subtitle from '../../components/Subtitle/Subtitle';
import {
  testIdHeader,
  testIdLoginButton,
  testIdLogoutButton,
  testIdToggleLanguageButton,
} from '../../constants/TestId';
import dummyAuth from '../../store/apis/dummy';
import { useAppDispatch, useAppSelector } from '../../store/hooks/useApp';
import { useFetchUserQuery } from '../../store/slices/auth/apis/dummyAuth';
import { selectCurrentUser, selectUserId } from '../../store/slices/auth/selectors';
import { logout } from '../../store/slices/auth/slice';
import { selectLocale } from '../../store/slices/translates/selectors';
import { toggleLocale } from '../../store/slices/translates/slice';

import type { ReactElement } from 'react';

import styles from './sass/Header.module.scss';

/**
 * Main layout header component
 */
function Header(): ReactElement {
  const { t } = useTranslation();
  const dispatch = useAppDispatch();
  const id = useAppSelector(selectUserId);
  const locale = useAppSelector(selectLocale);
  const { data, isFetching } = useFetchUserQuery(id, {
    skip: !id,
    selectFromResult: (result) => ({ ...result, data: selectCurrentUser(result.data, id) }),
  });
  const onLogout = () => {
    dispatch(dummyAuth.util.resetApiState());
    dispatch(logout());
  };
  return (
    <header className={styles.header} data-testid={testIdHeader}>
      <Subtitle text={t('common.welcomeMessage')} />
      <nav className={styles.headerNav}>
        <Link to="/" className={styles.headerNavLink}>
          { t('home.pageTitle') }
        </Link>
        <Link to="/about" className={styles.headerNavLink}>
          { t('about.pageTitle') }
        </Link>
      </nav>
      <div className={styles.toggleLanguage}>
        {isFetching ? (
          <Loader />
        ) : (
          <span>
            {data?.id ? (
              <LinkButton
                to="/login"
                onClick={onLogout}
                title={`${t('login.logout')}(${data.firstName})`}
                dataTestId={testIdLogoutButton}
              />
            ) : (
              <LinkButton
                to="/login"
                title={t('login.pageTitle')}
                dataTestId={testIdLoginButton}
              />
            )}
          </span>
        )}
        <button
          data-testid={testIdToggleLanguageButton}
          type="button"
          onClick={() => dispatch(toggleLocale())}
        >
          {locale.toUpperCase()}
        </button>
      </div>
    </header>
  );
}

export default Header;
