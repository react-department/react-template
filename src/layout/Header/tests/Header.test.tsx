import { fireEvent, screen, waitFor } from '@testing-library/react';

import { initServer } from '../../../../__tests__/mswServer';
import { render } from '../../../../__tests__/test-utils';
import {
  testIdLoader,
  testIdLoginButton,
  testIdLogoutButton,
  testIdToggleLanguageButton,
} from '../../../constants/TestId';
import auth from '../../../mocks/auth';
import Header from '../Header';

initServer();

describe('<Header />', () => {
  test('should render component', () => {
    const { asFragment } = render(<Header />);
    expect(asFragment()).toMatchSnapshot();
  });
  test('should contain needed elements', () => {
    render(<Header />);
    const title = screen.getByText('Try to build something great');
    const listLinks = screen.getAllByRole('link');
    const button = screen.getByTestId(testIdToggleLanguageButton);
    const loginLink = screen.getByText('Login');
    expect(title).toBeInTheDocument();
    expect(loginLink).toBeInTheDocument();
    expect(button).toBeInTheDocument();
    expect(listLinks).toHaveLength(3);
  });
  test('should change locale', async () => {
    render(<Header />);
    const button = screen.getByTestId(testIdToggleLanguageButton);
    fireEvent.click(button);
    await waitFor(() => {
      expect(button).toHaveTextContent('UA');
    });
  });
  test('should logout after click logout button', async () => {
    render(<Header />, { preloadedState: { auth } });
    await waitFor(() => expect(screen.queryByTestId(testIdLoader)).not.toBeInTheDocument());
    const button = screen.getByTestId(testIdLogoutButton);
    fireEvent.click(button);
    await waitFor(() => {
      const loginLink = screen.getByTestId(testIdLoginButton);
      expect(loginLink).toBeInTheDocument();
    });
  });
});
