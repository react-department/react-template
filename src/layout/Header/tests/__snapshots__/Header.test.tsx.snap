// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Header /> should render component 1`] = `
<DocumentFragment>
  <header
    class="header"
    data-testid="header"
  >
    <h2
      class="appSubtitle"
      data-testid="subtitle"
    >
      Try to build something great
    </h2>
    <nav
      class="headerNav"
    >
      <a
        class="headerNavLink"
        href="/"
      >
        Home
      </a>
      <a
        class="headerNavLink"
        href="/about"
      >
        About
      </a>
    </nav>
    <div
      class="toggleLanguage"
    >
      <span>
        <a
          class="linkButton"
          data-testid="loginButton"
          href="/login"
        >
          Login
        </a>
      </span>
      <button
        data-testid="toggleLanguageButton"
        type="button"
      >
        EN
      </button>
    </div>
  </header>
</DocumentFragment>
`;
